[clang-format] Fix crashes when the macro expansion is empty

amane-ame opened the pull request:
An empty expansion should be valid, like `echo 'A()' | clang-format --style='{Macros: [A(x)=x]}'`.

Fixes #119258.

amane-ame committed:
Fix crashes when the macro expansion is empty

owenca commented:
Please add unit tests.

amane-ame committed:
[clang-format] Add a testcase for empty macro

owenca commented:
You should add unit tests in `clang/unittests/Format/`.

amane-ame committed:
[clang-format] Move the testcase to unittest

owenca commented on clang/unittests/Format/MacroCallReconstructorTest.cpp:
```diff
@@ -217,6 +217,19 @@ TEST_F(MacroCallReconstructorTest, Identifier) {
 }
 
+TEST_F(MacroCallReconstructorTest, EmptyExpansion) {
+  auto Macros = createExpander({"A(x)=y"});
```
Changing `y` to `x` causes an assertion failure.

amane-ame committed:
[clang-format] Fix about Tokens in MacroCallReconstructorTest

When the macro expansion is empty, the tok::eof Token should be preserved.
The reason is we have marked the start point and end point on this token in MacroExpander::expand.

HazardyKnusperkeks commented on clang/lib/Format/MacroExpander.cpp:
```diff
@@ -233,6 +233,10 @@ MacroExpander::expand(FormatToken *ID,
     ++Result[0]->MacroCtx->StartOfExpansion;
     ++Result[Result.size() - 2]->MacroCtx->EndOfExpansion;
+  } else {
+    // If the macro expansion is empty, mark the start and end
```
```suggestion
    // If the macro expansion is empty, mark the start and end.
```

owenca reviewed (commented):
Please add a test case for https://github.com/llvm/llvm-project/issues/119258#issue-2728126922.

amane-ame committed:
[clang-format] Fix a comment

owenca commented on clang/test/Format/empty-macro.cpp:
```diff
@@ -0,0 +1,5 @@
+// RUN: grep -Ev "// *[A-Z-]+:" %s | clang-format -style="{Macros: [A(x)=x]}" \
+// RUN:   | FileCheck -strict-whitespace %s
+
+// CHECK: A()
+A()
```
Don't add a lit test. Instead, add a unittest similar to the  one you already added there.

amane-ame replied:
Sorry for misunderstood.

amane-ame committed:
[clang-format] Add a unittest for GH119258

owenca commented on clang/unittests/Format/MacroCallReconstructorTest.cpp:
```diff
@@ -217,6 +219,31 @@ TEST_F(MacroCallReconstructorTest, Identifier) {
   EXPECT_THAT(std::move(Unexp).takeResult(), matchesLine(line(U.consume("X"))));
 }
 
+TEST_F(MacroCallReconstructorTest, IdentifierObject) {
```
How about`EmptyDefinition` instead of `IdentifierObject`?

amane-ame committed:
[clang-format] Change unittest name
